flax/core/lift.py

```diff
@@ -872,37 +882,142 @@ def scanned(broadcast_vars, carry, scan_variable_groups, rng_groups, args):
       c, y = fn(scope, c, *args)
       out_vars = repack_fn(scope)
       broadcast_vars_out = out_vars[0]
-      carry_vars = out_vars[1]
+      carry_vars_out = out_vars[1]
       scan_vars = out_vars[2:]
+
+      # compute new carry vars, these will be handled as outputs
+      carry_vars_new = tuple(
+          vars_diff(outputs, inputs)
+          for outputs, inputs in zip(carry_vars_out, carry_vars)
+      )
+      # remove new carry vars to maintain input shape
+      carry_vars = tuple(
+          vars_diff(outputs, new)
+          for outputs, new in zip(carry_vars_out, carry_vars_new)
+      )
+
       # add immutable broadcast vars back to broadcast output
       # otherwise they won't be fed to the actual scan body
       for in_group, out_group in zip(broadcast_vars, broadcast_vars_out):
         for col in in_group:
           if col not in out_group:
             out_group[col] = in_group[col]
-      return broadcast_vars_out, (carry_vars, c), (y, scan_vars)
+      return broadcast_vars_out, (carry_vars, c), (y, scan_vars, carry_vars_new)
 
     broadcast_vars = variable_groups[0]
     carry_vars = variable_groups[1]
     scan_vars = variable_groups[2:]
     new_scan_vars = []
     for scan_group, axis in zip(scan_vars, variable_in_axes):
       new_scan_vars.append(meta.remove_axis(scan_group, axis, metadata_params))
-    broadcast_vars, (carry_vars, c), (ys, scan_vars) = scanned(
-        broadcast_vars,
-        (carry_vars, init),
-        tuple(new_scan_vars),
-        rng_groups,
-        args,
-    )
+
+    # compute new carry vars
+    with tabulate_context(
+        add_call_info=False
+    ):  # dont add call info while tracing
```

Comment:
The `add_call_info=False` option is added in this PR as a mechanism to turn tabulate entirely. Since we are calling the module multiple times, tabulate would register more calls, but we are only interested in the having one. Actually, I think I will change it so its the abstract one that add the tabulate info.